src/components/primitives/Text/Clamp.js

```diff
@@ -0,0 +1,149 @@
+import React, { PureComponent, Fragment } from "react";
+import { StyleSheet, Text, View } from "react-native";
+
+export default class Clamp extends PureComponent {
+  state = {
+    isTruncating: null,
+    containerWidth: null,
+    preEllipsisText: null,
+    restOfText: null,
+  };
+
+  handleSingleLineLayout = event => {
+    const singleLineWidth = event.nativeEvent.layout.width;
+    const {
+      numberOfLines,
+      children: text,
+      children: { length: totalCharCount },
+    } = this.props;
+    const { containerWidth } = this.state;
+
+    const averageCharsPerLine = getAverageCharsPerLine({
+      singleLineWidth,
+      totalCharCount,
+      containerWidth,
+    });
+
+    const { preEllipsisText, restOfText } = getText({
+      text,
+      numberOfLines,
+      averageCharsPerLine,
+    });
+
+    this.setState(state => ({
+      preEllipsisText,
+      restOfText,
+      isTruncating: false,
+    }));
+  };
+
+  handleContainerLayout = ({
+    nativeEvent: {
+      layout: { width },
+    },
+  }) => {
+    if (width === this.state.containerWidth) {
+      return;
+    }
+    this.setState(state => ({
+      containerWidth: width,
+      isTruncating: true,
+    }));
+  };
+
+  renderTextOnSingleLine = () => {
+    const { numberOfLines, style, ...rest } = this.props;
+    return (
+      <View style={styles.singleLineContainer}>
+        <Text
+          onLayout={this.handleSingleLineLayout}
```

Comment:
ah it looks like the compiler takes care of it.

Reply:
Actually I think that's a feature of fat arrows

Reply:
ah right because handleSingleLineLayout is defined as a class variable with a fat arrow it binds it there. That's pretty cool, but also kind of crazy that there is such a big semantic difference between that and a regular method.